### PR TITLE
extract code render function with query path string

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs [hunk 1]
```diff
@@ -32,44 +32,67 @@ public DfmExtractCodeResult ExtractFencesCode(DfmFencesToken token, string fence
...
+                Logger.LogWarning(GenerateErrorMessage(token), line: token.SourceInfo.LineNumber.ToString(), code: WarningCodes.Markdown.InvalidCodeSnippet);
+                return new DfmExtractCodeResult { IsSuccessful = false, ErrorMessage = token.PathQueryOption.ErrorMessage, CodeLines = fencesCode };
+            }
+            if (!string.IsNullOrEmpty(token.PathQueryOption.ErrorMessage))
+            {
+                Logger.LogWarning(GenerateErrorMessage(token), line: token.SourceInfo.LineNumber.ToString());
+            }
+
+            var includedLines = new List<string>();
+            foreach (var line in token.PathQueryOption.GetQueryLines(fencesCode))
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: `var includedLines = token.PathQueryOption.GetQueryLines(fencesCode).ToList()`?

File: src/Microsoft.DocAsCode.Dfm/DfmCodeRenderer.cs [hunk 2]
```diff
@@ -49,6 +73,11 @@ public virtual DfmExtractCodeResult ExtractCode(DfmFencesToken token, string fil
             return _dfmCodeExtractor.ExtractFencesCode(token, filePath);
         }
 
+        public virtual DfmExtractCodeResult ExtractCode(DfmFencesToken token, string[] fencesCode)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeRenderer.cs]: should we modify the parameters (pathQueryOption)?

File: src/Microsoft.DocAsCode.Dfm/DfmFencesBlockPathQueryOptions/TagNameBlockPathQueryOption.cs [hunk 3]
```diff
@@ -210,6 +210,21 @@ private DfmTagNameResolveResult ResolveTagNamesFromPath(string fencesPath, strin
...
 
+        private Lazy<ConcurrentDictionary<string, List<DfmTagNameResolveResult>>> GetLazyResolveResult(string[] fencesCodeLines, List<ICodeSnippetExtractor> codeSnippetExtractors)
+        {
+           return new Lazy<ConcurrentDictionary<string, List<DfmTagNameResolveResult>>>(
+                () =>
+                {
+                    // TODO: consider different code snippet representation with same name
+                    return new ConcurrentDictionary<string, List<DfmTagNameResolveResult>>(
+                        (from codeSnippetExtractor in codeSnippetExtractors
+                            let resolveResults = codeSnippetExtractor.GetAll(fencesCodeLines)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Dfm/DfmFencesBlockPathQueryOptions/TagNameBlockPathQueryOption.cs]: indent

File: src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs [hunk 4]
```diff
@@ -32,44 +32,63 @@ public DfmExtractCodeResult ExtractFencesCode(DfmFencesToken token, string fence
...
-                if (!token.PathQueryOption.ValidateAndPrepare(fencesCode, token))
-                {
-                    Logger.LogWarning(GenerateErrorMessage(token), line: token.SourceInfo.LineNumber.ToString(), code: WarningCodes.Markdown.InvalidCodeSnippet);
-                    return new DfmExtractCodeResult { IsSuccessful = false, ErrorMessage = token.PathQueryOption.ErrorMessage, CodeLines = fencesCode };
-                }
-                if (!string.IsNullOrEmpty(token.PathQueryOption.ErrorMessage))
-                {
-                    Logger.LogWarning(GenerateErrorMessage(token), line: token.SourceInfo.LineNumber.ToString());
-                }
+                return ExtractFencesCodeCore(token, fencesCode);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs]: >ExtractFencesCodeCore [](start = 23, length = 21)

call below ExtractFencesCode instead?

File: src/Microsoft.DocAsCode.Dfm/DfmFencesBlockPathQueryOptions/TagNameBlockPathQueryOption.cs [hunk 5]
```diff
@@ -60,6 +60,8 @@ public class TagNameBlockPathQueryOption : DfmFencesBlockPathQueryOption
 
         public string TagName { get; set; }
 
+        public bool NoCache { get; set; }
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Dfm/DfmFencesBlockPathQueryOptions/TagNameBlockPathQueryOption.cs]: >NoCache [](start = 20, length = 7)

pass in through ctor instead?

**Pull request merged**